packages/node-core/src/indexer/dynamic-ds.service.ts

```diff
@@ -91,6 +94,74 @@ export class DynamicDsService<DS extends BaseDataSource = BaseDataSource, P exte
     }
   }
 
+  getDynamicDatasourcesByTemplate(templateName: string): DynamicDatasourceInfo[] {
+    if (!this._datasourceParams) {
+      throw new Error('DynamicDsService has not been initialized');
+    }
+
+    const matchingDatasources = this._datasourceParams
+      .map((params, globalIndex) => ({params, globalIndex}))
+      .filter(({params}) => params.templateName === templateName && params.endBlock === undefined);
+
+    return matchingDatasources.map(({globalIndex, params}, index) => ({
+      index,
+      templateName: params.templateName,
+      startBlock: params.startBlock,
+      endBlock: params.endBlock,
+      args: params.args,
+    }));
+  }
+
+  async destroyDynamicDatasource(
+    templateName: string,
+    currentBlockHeight: number,
+    index: number,
+    tx?: Transaction
+  ): Promise<void> {
+    if (!this._datasources || !this._datasourceParams) {
+      throw new Error('DynamicDsService has not been initialized');
+    }
+
+    // Get all matching datasources using the existing method
+    const matchingDatasourcesInfo = this.getDynamicDatasourcesByTemplate(templateName);
+
+    if (matchingDatasourcesInfo.length === 0) {
+      throw new Error(`Dynamic datasource with template name "${templateName}" not found or already destroyed`);
+    }
+
+    // Validate index
+    if (index < 0 || index >= matchingDatasourcesInfo.length) {
+      throw new Error(
+        `Index ${index} is out of bounds. There are ${matchingDatasourcesInfo.length} active datasource(s) for template "${templateName}"`
+      );
+    }
+
+    // Get the datasource info at the specified index
+    const dsInfo = matchingDatasourcesInfo[index];
+
+    // Find the global index in _datasourceParams
+    const globalIndex = this._datasourceParams.findIndex(
+      (p) => p.templateName === dsInfo.templateName && p.startBlock === dsInfo.startBlock && p.endBlock === undefined
+    );
```

Comment:
Why not just have the index on `DynamicDatasourceInfo` then there's no need to filter by templateName, find the index, then find the global index.